### ISSUE TITLE
memory leak on long run (3-4 weeks) 

### ISSUE THREAD

**Issue by gkorland** [FalkorDB/FalkorDB#446]
Created by: @postpersonality
Source: https://github.com/RedisGraph/RedisGraph/issues/3103
Memory consumption of master redis-graph process.
![image](https://github.com/RedisGraph/RedisGraph/assets/2012215/9ce921d1-5b7d-4d7c-8d86-9fa1f3d60399)
(disregard dips at the start of the graph)

The master process receives only upsert queries (see below) at ~1.2k rps at peak and serves 2 replicas. Size of database dump is 46Mb. So, most of the time it updates existing nodes and edges.

At 9th June the DB throughput has degraded to 160 rps. At the same time memory graph started to creep upwards.

The query looks like
```
MERGE (af:Node {id: '${af}'}) ON CREATE SET af.id='${af}'
MERGE (at:Node {id: '${at}'}) ON CREATE SET at.id='${at}'
MERGE (af)-[f:Edge {id: '${id}'}]->(at)
ON MATCH SET
 f.pa=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN ${pa} ELSE f.pa END,
 f.ag=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN '${ag}' ELSE coalesce(f.ag, '') END,
 f.tx=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN ${tx} ELSE f.tx END
ON CREATE SET
 f.id='${id}',
 f.tx=${tx},
 f.ag='${ag}',
 f.pa=${pa}
```

There are indices
```
CREATE INDEX FOR (a:Node) ON (a.id)
CREATE INDEX FOR ()-[f:Edge]->() ON (f.id)
CREATE INDEX FOR ()-[f:Edge]->() ON (f.tx)
```

Redis version 6.2.6
RedisGraph version 2.8.26

**Comment by gkorland** [comment]
Origin comment by: @gkorland
Can you please try upgrade to 2.10?

**Comment by gkorland** [comment]
Origin comment by: @postpersonality
@gkorland is 2.10 stable enough to migrate? I've seen memory leaks fixes have been ported to 2.8.

**Comment by gkorland** [comment]
Origin comment by: @gkorland
Yes, 2.10.11 has many fixes on top of 2.8.

**Comment by gkorland** [comment]
Origin comment by: @postpersonality
@gkorland Is it backwards compatible? Will it read on-disk dump from 2.8 version?

**Comment by gkorland** [comment]
Origin comment by: @postpersonality
@swilly22 One thing to note if you are going to replicate the query. You may take in account that adding nodes is very-very rare case. On empty graph 99.99% of all nodes get settled in 3 days. But 95% of them get settled in hour. Maybe the trigger for the memory leak is the rare new node, just maybe. It's hard to know since the data stream is quite big.

Here is our current memory chart for redisgraph 2.12.8 master (upward slopes are the memory leaks)
![image](https://github.com/RedisGraph/RedisGraph/assets/2012215/1d332d3b-96d9-4ea7-8588-b462985d9a6f)

**Comment by gkorland** [comment]
Origin comment by: @gkorland
> @gkorland Is it backwards compatible? Will it read on-disk dump from 2.8 version?

Yes it's and it will read the dump

**Comment by postpersonality** [comment]
Hi, it seems with 2.12.9 it works quite well. There are no spikes in memory consumption, but still there are slow leaks. Currently it's around steady 13 mb per day.

![image](https://github.com/FalkorDB/FalkorDB/assets/2012215/427da0e4-8195-40bb-96ef-120b16659c0c)

The dip downs on the chart are DB reboots.
The linear slope around 17 Oct is missing data chart issue.

**Comment by postpersonality** [comment]
Unfortunately it's still leaking unexpectedly. Last restart of DB was on 17th Oct ()
![image](https://github.com/FalkorDB/FalkorDB/assets/2012215/850e9e1d-24c4-418c-ad2b-63c1264a819e)

**Comment by postpersonality** [comment]
We have noticed on our side that leaking starts when chef agent runs his check once an hour (the orange spikes). So, maybe that's a factor we need to take in account. 

![Screenshot from 2023-11-01 10-55-53](https://github.com/FalkorDB/FalkorDB/assets/2012215/ee9b11fb-5426-4ce2-b7c4-b832f21a1560)

![Screenshot from 2023-11-01 11-43-32](https://github.com/FalkorDB/FalkorDB/assets/2012215/907c7209-0936-4324-80d8-fdffd37baeee)

![Screenshot from 2023-11-01 11-00-00](https://github.com/FalkorDB/FalkorDB/assets/2012215/c55a4901-c661-4a46-b3a4-603e854a64cc)

**Comment by gkorland** [comment]
Please do share any info you might have, @swilly22 told me he'll try to get to the issue later

**Comment by swilly22** [comment]
Hi @postpersonality,
I've tried recreating the issue using the provided query:
```cypher
MERGE (af:Node {id: '${af}'}) ON CREATE SET af.id='${af}'
MERGE (at:Node {id: '${at}'}) ON CREATE SET at.id='${at}'
MERGE (af)-[f:Edge {id: '${id}'}]->(at)
ON MATCH SET
 f.pa=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN ${pa} ELSE f.pa END,
 f.ag=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN '${ag}' ELSE coalesce(f.ag, '') END,
 f.tx=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN ${tx} ELSE f.tx END
ON CREATE SET
 f.id='${id}',
 f.tx=${tx},
 f.ag='${ag}',
 f.pa=${pa}
```

Without much luck, let me suggest the following:
1. query rewrite:
```cypher
MERGE (af:Node {id: '${af}'})
MERGE (at:Node {id: '${at}'})
MERGE (af)-[f:Edge {id: '${id}'}]->(at)
ON MATCH SET
 f.pa=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN ${pa} ELSE f.pa END,
 f.ag=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN '${ag}' ELSE coalesce(f.ag, '') END,
 f.tx=CASE WHEN f.tx<${aggrTs} OR ${pa}<f.pa THEN ${tx} ELSE f.tx END
ON CREATE SET
 f.tx=${tx},
 f.ag='${ag}',
 f.pa=${pa}
```

As there's no need to re-set the `id` attribute when the entity is created.

2. Any additional information on the queries which hit the server when the memory spikes could help, this includes the output from GRAPH.SLOWLOG and a portion of output from captured MONITOR log.

3. If you like I'll be more than happy to join an online profiling session.

**Comment by postpersonality** [comment]
Well, original query is written to mimic all-in-one UPSERT query. 

As I understand your query 
1. It will not create nodes, I have to create them separately.
2. `MERGE (af)-[f:Edge {id: '${id}'}]->(at)` refers id property (the id property and id of the edge are different beasts) which is not created by your query

Yes, I know the property with name "id" is misleading. It should be called like "code" or something. This is basically string slug of nodes/edges to retrive them quickly by that slug (I have index for that). 

> As there's no need to re-set the id attribute when the entity is created.
That's correct. That's why we set id only in `ON CREATE` block which does exactly that as I understand.

**Comment by swilly22** [comment]
The rewrite I've proposed will have the exact same effect as the original query, when the MERGE pattern doesn't exists in the graph it will be created, as specified. setting attributes which are mentioned within the MERGE pattern will simply overwrite the attribute.

For example:
```cypher
MERGE (n:N {v:1})
```

Will create a new `N` node with the attribute `v` set to 1 if such node doesn't exists in the graph, 

```cypher
MERGE (n:N {v:1}) ON CREATE SET n.v = 1
```

Will do exactly the same thing but will waste addition time setting `v` to 1 a second time.

I hope this better explains how the `MERGE` clause works.

I would love to get more information regarding this leak, do you think we can take this online ?

**Comment by postpersonality** [comment]
Thanks for the detailed explanation. I have SQL background so that behavior is unexpected for me. Probably I missed that part in the Cypher docs.

> If you like I'll be more than happy to join an online profiling session.
Not sure if that's possible, I have no access to prod environment directly for security reasons.

But I'll try to make some dockerized emulation of our use-case and see if it can help reproduce the leak.

**Comment by swilly22** [comment]
Thank you for the additional details,
Can you please elaborate what do you mean by "internal RedisGraph cleanup cycle." ?

**Comment by postpersonality** [comment]
Good catch. That was my assumption. I forgot we have DELETE query to clean up old data once an hour.

So, the leak happens on (or close to) `MATCH ()-[f:Edge]->() WHERE f.ts<CONST_NUMBER_1 OR f.tx<CONST_NUMBER_2 DELETE f`.

**Comment by postpersonality** [comment]
By the way we still use unoptimized "upsert" query if it's related.

**Comment by swilly22** [comment]
To eliminate a potential candidate may I ask you apply the following configuration to your `master`:

```
GRAPH.CONFIG SET EFFECTS_THRESHOLD 0
```

This will force the `master` to always replicate queries using a compact and efficient way of representing graph modifications.
See [docs](https://docs.falkordb.com/configuration.html#effects_threshold)

Please let me know if you're still experiencing leaks on the `SLAVES` after the configuration changed been applied.

**Comment by swilly22** [comment]
Thank you,
Would it be possible to get a capture of the output from the `MONITOR` command running on the `MASTER`
```sh
redis-cli -h <MASTER_HOST> -p <MASTER_PORT> MONITOR > monitor.log
```

**Comment by postpersonality** [comment]
That's quite interesting. Today it's chill and no leaks. Many thanks, @swilly22!
![image](https://github.com/FalkorDB/FalkorDB/assets/2012215/1b1fe6bc-2b7f-4768-a0cb-b51131f5ba44)

**Comment by swilly22** [comment]
How can we explain this ?

**Comment by postpersonality** [comment]
Today is also no leaks. I think this is kind of config propagation error or faulty state of replicas which back to normal after their leak-oom-restart loop. We didn't restart replicas forcefully on the config change.

**Comment by swilly22** [comment]
One thing to note though is that the suggested configuration is only applicable to the `MASTER`.

**Comment by postpersonality** [comment]
Today a leak happened on replica that last time leaked before the config change and no restarts since then. My current observation is that after the config change there were no leaks on the same replica twice. For now it looks like replicas get fixed after 1) master config change 2) replica restart. 

Yes, I understand that the setting configures the master replication logic. But I assume replica state change, suppose it survived update command execution back in the past and no restarts since then. I'm sure I do not understand a thing how RedisGraph works, those are my assumptions.

Also please note the difference in memory consumption between replicas. All the replicas get more or less the same search queries and it seems they should be the same size including the search cache. Also master do not get any search queries if that matters.

**Comment by swilly22** [comment]
I understand.
In that case, I think I'll be interested in both a `MONITOR` log from the `master` and one from a leaking `replica` (if possible).

**Comment by postpersonality** [comment]
![image](https://github.com/FalkorDB/FalkorDB/assets/2012215/1ed14d15-9b2f-4a06-8211-4e617bbfeb0e)
The green is the leaking master. After that some replicas stopped to synchronize with "Partial resynchronization not possible (no cached master)"
We had to stop and start the whole cluster. And now memory consumption goes in line on all nodes.

**Comment by postpersonality** [comment]
I have some information about new case we had today at around 07:48/10:48 (logs with different timezones).
These are the logs from leaked master:
```
1248150:M 25 Dec 2023 07:43:20.760 * 10000 changes in 60 seconds. Saving...
1248150:M 25 Dec 2023 07:43:20.875 * Background saving started by pid 1416464
1416464:C 25 Dec 2023 07:43:20.877 * <graph> Created 57 virtual keys for graph graph:avia_routes
1248150:M 25 Dec 2023 07:43:50.761 # Can't fork for module: Success
1248150:M 25 Dec 2023 07:43:55.763 # Can't fork for module: Success
1416464:C 25 Dec 2023 07:43:55.947 * <graph> Done encoding graph graph:avia_routes
1416464:C 25 Dec 2023 07:43:56.091 * DB saved on disk
1416464:C 25 Dec 2023 07:43:56.092 * <graph> Deleted 57 virtual keys for graph graph:avia_routes
1416464:C 25 Dec 2023 07:43:56.227 * RDB: 1000 MB of memory used by copy-on-write
1248150:M 25 Dec 2023 07:43:56.475 * Background saving terminated with success
1248150:M 25 Dec 2023 07:45:04.087 * 10000 changes in 60 seconds. Saving...
1248150:M 25 Dec 2023 07:45:04.224 * Background saving started by pid 1416477
1416477:C 25 Dec 2023 07:45:04.227 * <graph> Created 57 virtual keys for graph graph:avia_routes
1248150:M 25 Dec 2023 07:45:34.087 # Can't fork for module: Success
1248150:M 25 Dec 2023 07:45:39.089 # Can't fork for module: Success
1416477:C 25 Dec 2023 07:45:39.563 * <graph> Done encoding graph graph:avia_routes
1416477:C 25 Dec 2023 07:45:39.731 * DB saved on disk
1416477:C 25 Dec 2023 07:45:39.732 * <graph> Deleted 57 virtual keys for graph graph:avia_routes
1416477:C 25 Dec 2023 07:45:39.868 * RDB: 1040 MB of memory used by copy-on-write
1248150:M 25 Dec 2023 07:45:40.112 * Background saving terminated with success
1248150:M 25 Dec 2023 07:46:46.242 * 10000 changes in 60 seconds. Saving...
1248150:M 25 Dec 2023 07:46:46.341 * Background saving started by pid 1416493
1416493:C 25 Dec 2023 07:46:46.344 * <graph> Created 57 virtual keys for graph graph:avia_routes
1248150:M 25 Dec 2023 07:47:16.242 # Can't fork for module: Success
1248150:M 25 Dec 2023 07:47:21.244 # Can't fork for module: Success
1416493:C 25 Dec 2023 07:47:22.522 * <graph> Done encoding graph graph:avia_routes
1416493:C 25 Dec 2023 07:47:22.661 * DB saved on disk
1416493:C 25 Dec 2023 07:47:22.661 * <graph> Deleted 57 virtual keys for graph graph:avia_routes
1416493:C 25 Dec 2023 07:47:22.798 * RDB: 997 MB of memory used by copy-on-write
1248150:M 25 Dec 2023 07:47:23.065 * Background saving terminated with success
1248150:M 25 Dec 2023 07:57:19.061 # Connection with replica 127.0.0.1:56400 lost.
1248150:M 25 Dec 2023 07:57:20.181 * Replica 127.0.0.1:56400 asks for synchronization
1248150:M 25 Dec 2023 07:57:20.182 * Partial resynchronization request from 127.0.0.1:56400 accepted. Sending 95160 bytes of backlog starting from offset 26054174424.
1248150:M 25 Dec 2023 07:58:34.326 # Connection with replica client id #2738 lost.
1248150:M 25 Dec 2023 07:58:34.985 * Replica 127.0.0.1:56400 asks for synchronization
1248150:M 25 Dec 2023 07:58:34.985 * Partial resynchronization request from 127.0.0.1:56400 accepted. Sending 61072 bytes of backlog starting from offset 26060326305.
1248150:M 25 Dec 2023 07:58:58.977 # Connection with replica client id #2723 lost.
1248150:M 25 Dec 2023 07:58:59.847 * Replica 127.0.0.1:56400 asks for synchronization
1248150:M 25 Dec 2023 07:58:59.847 * Partial resynchronization request from 127.0.0.1:56400 accepted. Sending 78720 bytes of backlog starting from offset 26062215259.
1248150:M 25 Dec 2023 07:59:08.934 # Connection with replica client id #2720 lost.
1248150:M 25 Dec 2023 07:59:09.624 * Replica 127.0.0.1:56400 asks for synchronization
1248150:M 25 Dec 2023 07:59:09.624 * Partial resynchronization request from 127.0.0.1:56400 accepted. Sending 50813 bytes of backlog starting from offset 26062925239.
1248150:M 25 Dec 2023 08:00:00.813 # Connection with replica client id #2746 lost.
1248150:M 25 Dec 2023 08:00:01.042 * Replica 127.0.0.1:56400 asks for synchronization
1248150:M 25 Dec 2023 08:00:01.043 * Partial resynchronization request from 127.0.0.1:56400 accepted. Sending 16407 bytes of backlog starting from offset 26066808607.
1248150:M 25 Dec 2023 08:00:45.738 # Connection with replica client id #2748 lost.
1248150:M 25 Dec 2023 08:00:46.687 * Replica 127.0.0.1:56400 asks for synchronization
1248150:M 25 Dec 2023 08:00:46.687 * Partial resynchronization request from 127.0.0.1:56400 accepted. Sending 98963 bytes of backlog starting from offset 26070241743.
```

After that time DB snapshotting has stopped working. See the I/O Activity:
![image](https://github.com/FalkorDB/FalkorDB/assets/2012215/708b9378-80a0-470f-b171-d036d5fde332)

At the same time memory consumption profile has changed:
![image](https://github.com/FalkorDB/FalkorDB/assets/2012215/4d36f3b8-5477-4dec-a00b-045003806ac5)

By the way we do not have replica (127.0.0.1:56400) located on the master. I guess that's snapshotting mechanism.

**Comment by postpersonality** [comment]
I have tried to make dockerized environment with our use case. And I did it, but not succeeded finding the leak in 2 hours of continuous run.

**Comment by postpersonality** [comment]
We've noticed that frequency of the leaks has lowered significantly when we have reduced flow of write queries and as a consequence the size of the database in memory has reduced. The leaks on replicas has stopped.

One thing to note. That the size of redis in memory was close to the VM limit of 6Gb, the size was around 5-4Gb, now it's 1.5Gb. I've tried to replicate the case with docker container limits but with no luck. Maybe I should try to emulate the case with VMs.
